distributed/tests/test_worker_state_machine.py

```diff
@@ -1754,3 +1755,18 @@ def assert_time(actual, expect):
     assert_time(elapsed["x", "cancelled"], 0.15)
     assert_time(elapsed["y", "waiting"], 0.1)
     assert_time(elapsed["z", "flight"], 0.25)
+
+
+def test_remove_worker(ws):
+    ws.handle_stimulus(
+        ComputeTaskEvent.dummy(
+            key="z",
+            who_has={"y": ("tcp://127.0.0.7:1", "tcp://127.0.0.7:2")},
+            priority=(0,),
+            stimulus_id="compute2",
+        ),
+        RemoveWorkerEvent(stimulus_id="remove1", worker="tcp://127.0.0.7:2"),
+    )
+    assert ws.data_needed == {}
+    assert ws.tasks["y"].who_has == {"tcp://127.0.0.7:1"}
+    assert ws.has_what == {"tcp://127.0.0.7:1": {"y"}}
```

Comment:
```suggestion
    ws2 = "127.0.0.1:2"
    ws3 = "127.0.0.1:3"
    ws.handle_stimulus(
        ComputeTaskEvent.dummy("z", who_has={"y": [ws2, ws3]}, stimulus_id="s1"),
        RemoveWorkerEvent(worker=ws3, stimulus_id="s2"),
    )
    assert not ws.data_needed
    assert ws.tasks["y"].who_has == {ws2}
    assert ws.has_what == {ws2: {"y"}}
```

Reply:
Between the two stimuli, could you add `assert ts.coming_from == ws3`?

Reply:
The pick of worker to gather from is pseudo-random. I made this test more robust.